### ISSUE TITLE
credocument doesn't implement browse by bookmarks

### ISSUE THREAD

**Issue by ersi-dnd** [koreader/koreader#2819]
* KOReader version:
* Device:

#### Issue
This is an enhancement suggestion.

My suggestion concerns the behaviour of Next/Previous functions. I'd like to be able to browse by going to next/previous bookmark/TOC item (in my mind they are the same things anyway). 

Long press in the upper right or upper left corner in a pdf/djvu file brings up some sort of overview mode. 
In my opinion, my suggestion would fit neatly under a long press in the upper right corner, while a long press in the upper left corner should remain as it is.

So the suggestion is this: Long press in the upper right corner should activate the function that makes Next/Previous browse by bookmarks and TOC items instead of by pages. This should work for all file types. This browsing mode should be indicated by a little icon in the upper right corner, different from the bookmark icon. When there are no bookmarks or TOC items to browse, Next/Previous  should go to the very end/beginning of the file.

**Comment by Frenzie** [comment]
> my suggestion would fit neatly under a long press in the upper right corner

That's exactly how it works actually, but it's currently only implemented for PDF and DjVu. https://github.com/koreader/koreader/wiki/Tips-and-Tricks#2-bookmark-browsing-mode-only-for-pdf-and-djvu (okay, and it only does bookmarks, not TOC which is very much not the same thing in my mind — my main use is switching between endnotes and the text).

**Comment by ersi-dnd** [comment]
Question 1. So because of this thing called crengine, it cannot be implemented in epubs and other text files?

Question 2. > one of my uses is switching between endnotes and the text I'm currently reading
This use case would be covered by switching back and forth between any arbitrary bookmarks/TOC items, which are accessible in the bookmarks/TOC view. It's of course tougher to hit at the right items in the bookmarks/TOC view and easier to designate two arbitrary points in the file and then switch back and forth between those by means of Next/Previous. 

The reason I brought this up is that Pocketbook has nice double functions for its keys: Push is one action and Long press another, and those actions are configurable to an extent. This does not work in Koreader, where Next/Previous keys are just Next/Previous, so I'm thinking of ways/workarounds to give Next/Previous keys more to do in Koreader.

**Comment by Frenzie** [comment]
> Question 1. So because of this thing called crengine, it cannot be implemented in epubs and other text files?

Not cannot, but it's entirely separate from the PDF interface (as is the DjVu one).

> This use case would be covered by switching back and forth between any arbitrary bookmarks/TOC items

Naturally, that's actually mainly how I do it, but my point is that when the next TOC marker is mere pages away and the next bookmark hundreds, they are very much not the same thing.

**Comment by ersi-dnd** [comment]
> Naturally, that's actually mainly how I do it, but my point is that when the next TOC marker is mere pages away and the next bookmark hundreds, they are very much not the same thing.

True. So bookmarks are like marks (as in `vi` and the like) and TOC is different, but should this mean that TOC items should not be optionally browsable by Next/Previous? Is there such a mode or option?

**Comment by ersi-dnd** [comment]
As to browsing by bookmarks in pdf, this seems to work only by flip or swipe gesture (or whatever you call that gesture). An aspect of my suggestion was to make this work by Next/Previous keys. In Pocketbook it doesn't work with the keys. 

On e-readers, I strongly prefer the physical keys over touchscreen. I'd like to do things by means of keys as far as possible. Kobo lacks keys, one of the major reasons why I never felt I should get me a Kobo.

**Comment by poire-z** [comment]
(Some precision about question1: crengine is the rendering engine that deals with epubs and txt document (so, these are credocuments) - it's what need to be modified to implement browse by bookmarks, thus the issue's title change by Frenzie).

I didn't even know this flipping mode and browse by bookmarks, just tested it today.
But the nice thing about this working only with swipe is that, when you jump with swipe to a bookmark, you can browse the next/prev pages with tap (to continue reading a parapraph at the end of the bookmarked page), and then use swipe again to go to next bookmark.
If we make tap follows next bookmarks (like swipe does), one need to toggle off flipping mode to see what's on the next page, which is less convenient.

About browse by TOC/chapters, I remember @frenzie talked some weeks ago (when talking about swipe down for menu) about implementing it with swipe east/west at the bottom of the screen.
(But please don't use bottom ! while holding the device with one hand at the bottom, depending on position/balance, I sometime use swipe instead of tap to go to next/prev page - taping would make me lose grip for some milliseconds:)

I actually have this in my personnal patch, in readerrolling/readerpeging:
```lua
     self.ui:registerTouchZones({
        { -- more functions available when swiping on top
            id = "paging_top_swipe",
            overrides = {'paging_swipe'},
            ges = "swipe",
            screen_zone = {
                ratio_x = 0, ratio_y = 0, ratio_w = 1, ratio_h = 1/4,
            },
            handler = function(ges) return self:onTopSwipe(nil, ges) end
        },
```

```lua
-- more functions available when swiping on top
function ReaderPaging:onTopSwipe(_, ges)
    if ges.direction == "west" then
        local pageno = self.current_page
        self:onGotoPage(self.ui.toc:getNextChapter(pageno, 0))
    elseif ges.direction == "east" then
        local pageno = self.current_page
        self:onGotoPage(self.ui.toc:getPreviousChapter(pageno, 0))
    elseif ges.direction == "south" then
        -- cause my power button is getting loose, and I don't want to make it worse
        UIManager:scheduleIn(0.5, UIManager.event_handlers["Suspend"])
    elseif ges.direction == "north" then
        -- some custom stuff I added to kobo/powerd.lua to cycle thru predefined hardcoded light levels
        -- (that was before kobofrontlight plugin was introduced)
        local powerd = Device:getPowerDevice()
        local newintensity = powerd:cycleFrontlight()
        G_reader_settings:saveSetting("frontlight_intensity", newintensity)
        local Notification = require("ui/widget/notification")
        UIManager:show(Notification:new{
            text = "Frontlight intensity cycled to : "..newintensity,
            timeout = 0.3,
        })
    end
    return True
end
```
So, if there's some interest, and if we can assume people don't usually swipe left/right on top to change pages, that could be added (with other function for north swipe - the south swipe is actually overriden by the menu swipe :) in the middle, but works on the sides - but there may be some conflict/side effect with swipe north or south vs panning a document, so not sure about north).
(I could still hack to that my own needed stuff for south and north.)

**Comment by Frenzie** [comment]
@poire-z I don't think we should overload swiping with too much stuff by default but as I indicated in https://github.com/koreader/koreader/pull/3248#issuecomment-331620467 I think the likes of next/previous bookmark/chapter could make for great additions to the Skim to widget.

**Comment by poire-z** [comment]
I rarely use that widget, and i hardly have any bookmark, so can't say much what I'd like to find here.
And that widget will have chances be on top of the highlighted bookmarked text, so: close, read bookmarked text, relaunch that widget. Not my favorite tool :)
But if you add buttons for first/prev/next/last bookmarks, why not add another line of buttons for first/prev/next/last chapters :)

**Comment by Frenzie** [comment]
That's what I'm saying. I just hadn't thought of chapters when I slapped that together.

**Comment by ersi-dnd** [comment]
@Frenzie
>  I think the likes of next/previous bookmark/chapter could make for great additions to the Skim to widget.

Agreed. And looks like a sensible mockup. As the Skim widget currently is, I have no use for it.

**Comment by Frenzie** [comment]
Effectively implemented in #3389 through a different dialog.
